Issue#23: Intergrate user system

rbretmounet opened the pull request:
#23
### Summary:
 - Created a MySQL database on AWS which is then used to store user id, username, and password. Added the ability to create accounts and login with a password, delete a user's account, and change a user's password by typing in their old and new password. 
### Important Notes:
- Added 2 new packages for accessing the DB that has been added to the "requirements.txt" file. (Flask-SQLAlchemy and PyMySQL)
- Webapp currently accesses the DB through a user admin account which has permissions only required for reading and writing of user data. (Making it more secure)
- Access to the MySQL database is not very secure (AKA don't store any personal information that you don't want to be obtained.)
- Currently, have an admin account set up for login in. (Has no special permissions currently.)

### Future Updates:
- Make database more secure by having exclusive IPs that can access the database and have an initial config file for DB username, password, and IP that isn't stored on Github. (Doesn't have to be done till we release web app to the public or private info is stored on the DB.)
- Add other features such as admin privileges, account customization such as profile pics, etc.

Commit:
Merge branch 'master' of https://github.com/DVC-Viking-Robotics/webapp into DEV_Branch

Commit:
Merge branch 'master' of https://github.com/DVC-Viking-Robotics/webapp into issue23-intergrate-user-system

Commit:
added a password box for saving password when creating an account.

Commit:
added a password box for saving password when creating an account.

Commit:
Merge branch 'master' of https://github.com/DVC-Viking-Robotics/webapp into issue23-intergrate-user-system

Commit:
Merge branch 'master' of https://github.com/DVC-Viking-Robotics/webapp into issue23-intergrate-user-system

Commit:
changed code to store new users into a mysql database

Commit:
Made it so users can login with the store user accounts on the database.

Commit:
Merge branch 'master' of https://github.com/DVC-Viking-Robotics/webapp into issue23-intergrate-user-system

Commit:
added two new packages required for database

Commit:
added f to print fucntion for localhost

Commit:
added a / redirect at the begging to remove must login notification on first load

Commit:
changed database from localhost connection to a AWS database for dev.

Commit:
added the ability to delete account from database

Commit:
added a flash for saying account has been deleted

Commit:
added condition for when account already exists.

Commit:
changed color of flash messages based on succes and error

Commit:
changed color of flash messages based on succes and error

Commit:
added bootbox for alerts.

Commit:
removed bootbox

Commit:
added prompt asking user if they are sure about deleting account.

Commit:
fixed the delete account button looking funny

Commit:
changed modal name for delete user prompt

Commit:
added ability to change password

Commit:
moved all modals out of nav bar

tejashah88 commented on webapp/app.py:
```diff
@@ -9,11 +9,10 @@
 # to temporarily disable non-crucial pylint errors in conformity
 # pylint: disable=invalid-name,missing-docstring,no-value-for-parameter
 
-app = Flask(__name__)
```
We should have this variable in `app.py`. It's not obvious as to why the `app` variable is coming from `users.py`. If you look [slightly below](https://github.com/DVC-Viking-Robotics/webapp/blob/master/webapp/app.py#L23), you'll see that instead, there are `init_app` functions that modify `app`. I'd recommend to make an `init_app` function that does the same thing.

2bndy5 replied:
There's no need to write a `init_app()` function because [flask_sqlalchemy.SQLAlchemy already has one](https://flask-sqlalchemy.palletsprojects.com/en/2.x/api/#configuration), so I agree. Change it to `db = SQLAlchemy()`, put the `app` variable (along with the moved `login_manager.init_app(app)` line) back into the `app.py` file and import db into `app.py`. Then you can use `db.init_app(app)` in `app.py`. The `app.config["key"] = value` stuff should also live in `app.py`. This may seem trivial, but it is essential to keep orgaization of the code conformitive, otherwise we'll be looking at yet another refactoring crusade over the winter break (not productive).

2bndy5 commented on webapp/users.py:
```diff
@@ -18,33 +34,31 @@ def __init__(self, name, link='/remote'):
+        return False
+ 
     def get_id(self):
-        """This class attrubute holds the user account's ID"""
```
Publicly available functions need a docstring. Please add this back in... You may need to adjust what the docstring says since the `self._id` variable seems to hold different info now.

rbretmounet replied:
I believe that function is still in there I just have two of the same functions so I removed the duplicate on.

rbretmounet replied:
def get_id(self):
        """This class attrubute holds the user account's ID"""
        return str(self.id)

this is the function I have now for some reason github cut it out.

2bndy5 commented on webapp/users.py:
```diff
@@ -62,10 +76,10 @@ def get_id(self):
         return self._id
 
-users = {'admin': User(u'admin'), 'anonymous': AnonUser()}
+users = {'admin': User(u'admin','"admin'), 'anonymous': AnonUser()}
```
We can probably get rid of the `AnonUser` class and remove the `'anonymous': AnonUser()` user from this pre-existing users list. That is more of an artifact that I was playing around with when I [half-assed] setup the user system.

Commit:
removed users variable since it is no longer needed.

2bndy5 commented on webapp/users.py:
```diff
@@ -18,33 +34,31 @@ def __init__(self, name, link='/remote'):
+    def __repr__(self):
+        return '<User %r>' % (self.username)
+        
 class AnonUser(AnonymousUserMixin):
```
We don't need this class. Everything should require a login now

2bndy5 commented on webapp/routes.py:
```diff
@@ -2,37 +2,53 @@
+from sqlalchemy.sql import exists
+from flask import Blueprint, render_template, request, flash, redirect, session
+from flask_login import login_required, login_user, logout_user, current_user
+from .users import users, User, db
```
the `users` variable no longer exists

Commit:
removed unnecessarry function and import.

tejashah88 commented on webapp/routes.py:
```diff
@@ -109,3 +125,33 @@ def shutdown_robot():
+@blueprint.route("/delete_user")
+@login_required
+def delete_user():
+    "Deletes users account"
```
Make it surrounded by triple quotes to make it a docstring.

tejashah88 commented on webapp/users.py:
```diff
@@ -3,9 +3,25 @@
+app = Flask(__name__)
+
+# put them all together as a string that shows SQLAlchemy where the database is
+app.config['SQLALCHEMY_DATABASE_URI'] = 'mysql+pymysql://UserAdmin:^0o9JpRAPazf@webapp-dev-db.c1cjueiaoepn.us-west-1.rds.amazonaws.com:3306/user.accounts'
```
Let's put this string as a server config option, or at least not hardcoded (??). I don't think that it's secure to leave DB strings out in the wild.

tejashah88 commented on webapp/users.py:
```diff
@@ -3,9 +3,25 @@
+from flask_sqlalchemy import SQLAlchemy
 from flask_login import UserMixin, AnonymousUserMixin, LoginManager
 
+app = Flask(__name__)
```
See above comment about `app` variable

tejashah88 commented on webapp/users.py:
```diff
@@ -18,32 +34,28 @@ def __init__(self, name, link='/remote'):
+    __tablename__ = 'users'
+    id = db.Column('user_id',db.Integer, primary_key=True)
+    username = db.Column('username', db.String(20), unique=True, index=True)
+    password = db.Column('password', db.String(10))
```
We should be hashing the passwords for security sake. I'd recommend to look at [this article](https://pythonprogramming.net/password-hashing-flask-tutorial/) for more info.

tejashah88 commented on webapp/routes.py:
```diff
@@ -109,3 +125,33 @@ def shutdown_robot():
+
+@blueprint.route("/reset_password", methods=['GET', 'POST'])
+@login_required
+def reset_password():
```
Password also should be hashed here as well.

tejashah88 reviewed (changes requested):
Good work so far. I put a few comments about password hashing and where to define the `app` variable. Also, the `CREATE` and `grant` seem to be made by accident, so please delete them unless it's needed for some reason.

tejashah88 commented:
FYI, I will double check the AWS configuration and security settings before merging. @2bndy5 @rbretmounet

Commit:
changed to docstring

2bndy5 commented:
I'm also getting this error, but I don't know what it means...
![image](https://user-images.githubusercontent.com/14963867/66688462-92b3e700-ec3b-11e9-8db1-2a93d183945b.png)

2bndy5 commented:
@rbretmounet It may seem like we're tearing your work apart, but we are very excited to get this feature integrated and putting this issue behind us. Thank you for your commitment to this issue!

2bndy5 committed:
password input now has a key icon

2bndy5 committed:
removed another problematic syntax in login.html

2bndy5 commented:
Should we have the new user automatically logged in after successful account creation? When I created an account the "sign-up" modal just disappeared forcing me to login with the new account.

tejashah88 commented:
@2bndy5 I'd say to leave it as-is, so that the user is forced to enter (and thus remember) the password

tejashah88 commented:
I agree with @2bndy5, we are quite grateful for the work you put in, but we want to make sure this is polished up and ready to go.

Commit:
refractored code so all intiation is in app.py

Commit:
refractored code so everything is initialized in app.py

Commit:
Merge branch 'issue23-intergrate-user-system' of https://github.com/DVC-Viking-Robotics/webapp into issue23-intergrate-user-system

rbretmounet commented:
@2bndy5 and @tejashah88 I understand I have no prior experience with SQL databases and not a lot with python in general so I expect comments and mistakes. And I made it flash an account created message and redirect to login to retype but it is possible to just change the redirect to home after account creation.

Commit:
added password hashing for increased security

Commit:
Removed URI string and instead replaced with encrypted URI

Commit:
added cryptography as a rerquired package

2bndy5 committed:
deprecated anonymous user class

2bndy5 commented:
![image](https://user-images.githubusercontent.com/14963867/66696165-6e2d2e80-ec7e-11e9-84f5-4fdb1975bdac.png)
looks like the flash messages are rendering behind the navbar and are no longer fading?!

2bndy5 commented:
I think the `output.encrypted` file should be gitignored. Although I don't know if it should exist in the repo... My test run modified the file and now wants to stash it to the next commit. Also I'm deleting the files `CREATE` and `grant` per @tejashah88 suggestion. I don't know what any of these files are doing, so I'm waiting for input from you guys before making a commit.

2bndy5 committed:
removed `CREATE` & `grant` files

2bndy5 committed:
about page's deps list removals

2bndy5 committed:
some tiny twaeks

- del old image files related to old favicon
- fixed a syntactical error in layout.html
- re-order imports in app.py

rbretmounet commented:
@2bndy5 The `output.encrypted` file is the server string that is used to connect to the MySQL DB it changes every time because when the webapp in ran the file is decrypted into  `server_config.txt` then read and stored into `URI` and after that, the file closes and encrypts it. So it is crucial to have in the repo cause without it you won't be able to connect to the DB.

Commit:
changed reset_password so it works with the password hashing

Commit:
added output.encrypted to gitignore

Commit:
added .DS_Store to gitignore

Commit:
removed .DS_Store

Commit:
seeing if output.encrypted is still being pushed

tejashah88 commented:
It's the `key.key` file that needs to be gitignored, otherwise our whole encryption crusade is useless if someone can use that key. I'll generate a new key just to be safe and remove the old one.

tejashah88 committed:
cleaning up HTML and crypto-related files

tejashah88 committed:
made ID as indexable key

tejashah88 committed:
simplified managing fernet encrypted files

tejashah88 committed:
Added key recovery generator script in case of compromised keys

tejashah88 committed:
we don't need to re-encrypt the db config unless compromised

tejashah88 committed:
utf-8 decoding fix

tejashah88 committed:
temporary fix for disabling DB for testing purposes

tejashah88 committed:
changed to DISABLE_AUTH_SYSTEM flag name
